### PR TITLE
Vertical modes

### PR DESCRIPTION
**Pull request by crocha700**
This implements simple methods to compute vertical modes and project solutions onto those modes. The [3-layer example](http://git.io/vlfrv) was updated to showcase the usage of these methods. Unit tests are implemented.

### CODE REVIEW

File: docs/examples/layered.rst [hunk 1]
```diff
@@ -167,6 +167,56 @@ A snapshot and some diagnostics
...
+
+.. code:: python
+
+    m.vertical_modes()
+
+We can also project the solution onto the modes
+
+.. code:: python
+
+    pn = m.modal_projection(m.p)
```
Comment (rabernat) [inline docs/examples/layered.rst]: This is super cool! I like this way of doing it.

File: pyqg/model.py [hunk 2]
```diff
@@ -205,7 +205,41 @@ def run(self):
...
+        self.radii = np.zeros(self.nz)
+        self.radii[0] = self.g*self.H/np.abs(self.f) # barotropic def. radius
+        self.radii[1:] = np.sqrt(1./evals[asort][1:])
+
+        # eigenstructure 
+        self.pmodes = evecs[:,asort]
+
+        # normalize to have unit L2-norm
+        Ai = (self.H / (self.Hi[:,np.newaxis]*(self.pmodes**2)).sum(axis=0))**0.5
+        self.pmodes = Ai[np.newaxis,:]*self.pmodes
```
Comment (rabernat) [inline pyqg/model.py]: This code looks super clean. I love it.

Comment (rabernat) [inline pyqg/model.py]: Could you add ``radii``  and ``pmodes`` to the class attributes docstring (line 60 of model.py)?

**Comment by rabernat** [comment]
This is a great PR.

**Comment by francispoulin** [comment]
@crocha700 this looks great!  A few very minor comments.

You define kd = 1/Ld to be the inverse wavenumber.  Technically, I think you should have a 2*pi in there.  Not to say that this distracts from the coolness of your results but I thought I would mention it.

Also, is there a paper you could cite?  I know that Glenn has an earlier paper for the 2-layer vs 2-mode that shows they are equivalent, more or less.  But the 3-layer/3-mode problem is richer and if we pointed the reader (maybe me ) to the paper that might help.

Finally, you compute the kinetic energy density in terms of layers.  If you have a modal decomposition, I presume  you could easily do that in terms of the vertical modes.  That might be interesting to see.

Great job!

**Comment by crocha700** [comment]
Thanks for the reviews guys!

@francispoulin I  followed Glenn and many others, and defined the deformation radii as the square root of the inverse eigenvalue. I believe that that's a confusion that exists our field...To try to avoid this confusion, some (including Bill and I) call L_d = 2*pi/k_d the deformation length, as opposed to deformation radius  R_d = 1/k_d.  Just updated the docs with a citation to [Flierl 1978](http://www.sciencedirect.com/science/article/pii/0377026578900027).

As for the energy, I agree that you can define it modal wise energy, but since we're doing the advection in physical space (as opposed to modal space), there's no point in doing this online. The inclusion  of a modal subclass would be interesting, but that's a matter for a different  PR.

**Comment by rabernat** [comment]
> The inclusion of a modal subclass

You mean one that actually solves the modal equations? Interesting idea.

**Comment by rabernat** [comment]
> but since we're doing the advection in physical space (as opposed to modal space), there's no point in doing this online

I think that online diagnostics of modal KE would actually be very useful. This breakdown can't be recovered from the depth-space average, due to the coupling between modes. But that doesn't have to go in this PR.

**Comment by crocha700** [comment]
Let me try to implement these diagnostics then (don't merge this yet). It is actually quite simple: the modal energy spectrum is simply

(wv2+kn^2)*|ph_n|^2, where ph_n is the projection of ph onto the nth mode.

**Comment by crocha700** [comment]
>You mean one that actually solves the modal equations? Interesting idea.

This is the idea that Gleen introduced and [Hua and Hadivogel 1986](http://journals.ametsoc.org/doi/abs/10.1175/1520-0469(1986)043%3C2923%3ANSOTVS%3E2.0.CO%3B2) implemented. Andy and Ross also used a similar approaches for their theses (though they restricted themselves to the two-mode case). There are some advantages to the dynamics: Gleen argues that for "realistic" simulations one doesn't need to "calibrate" the modal mode (as opposed to matching deformation radii and interaction coefficients for layer models, etc.). [We've recently generalized this approach to correctly account for inhomogeneous  boundary conditions](http://git.io/vlOtO), but I somehow arbitrarily decided to pursue a highly accurate finite differences/Pade approximation in the vertical and never implemented those equations...Might be a good excuse to do it now.

**Comment by crocha700** [comment]
OK, I included  modal KE and PE spectra to diagnostics, and updated the example.

**Comment by francispoulin** [comment]
@crocha700 Thanks for clarifying about the radii vs length.  I see your point and will try and follow suit.

Also, thanks a lot for updating the example.  I'm glad it wasn't too much trouble and I do think that it's interesting.  It shows that at the large scales we have most of the energy is in the BT mode as we should expect.  Plotting the PE is also a nice touch.

Tomorrow I going to try and update my pyqg.  I haven't done it in a while and it's going through some big changes and I'm excited to see how things work.  I asked a question a while ago about including the linear stability analysis of the Bickley let.  I would like to include that and even do a multi-layer version.  True that we probably won't go higher than 4 layers or so but still, that would be interesting.

I also would like to do nonlinear simulations with pyqg.  Since the Bickley jet does not have periodic BCs, I think I would need to decompose the flow into a basic state and a perturbation and solve the nonlinear perturbation equations.  I know the equations, I'm just not sure how this will fit in the pyqg framework.

I realize this is not really part of issue 108 and maybe I should have started another one.  Sorry about that.

**Comment by rabernat** [comment]
My only questions is whether we should be plotting isotropic spectra in the example.

**Comment by crocha700** [comment]
OK, I updated the [example](http://git.io/vlfrv) to plot isotropic spectra. (I'll submit a different PR updating, testing and documenting those diagnostic_tools methods)

**Comment by crocha700** [comment]
@francispoulin let's move the Bickley jet discussion to #71

**Comment by rabernat** [comment]
Perfect!

**Pull request merged**